### PR TITLE
Add field level

### PR DESCRIPTION
**Pull request by MarKoz162**
Add `level` field to recipes & return it in response. (L)
And repair video_link validations.

### CODE REVIEW

File: app/models/recipe.rb [hunk 1]
```diff
@@ -1,8 +1,15 @@
 class Recipe < ApplicationRecord
   validates :name, presence: true
-  validate :must_link_be_correct 
+  validate :must_link_be_correct,:must_level_be_correct
+  
+  def must_level_be_correct
+    return unless level
+    errors.add(:level,"the number must be between 1 and 5") unless (1..5).include? level
+  end
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:8]: you should use rails validation `inclusion` instead of creating your own.

File: app/models/recipe.rb [hunk 2]
```diff
@@ -1,8 +1,15 @@
...
+  def must_level_be_correct
+    return unless level
+    errors.add(:level,"the number must be between 1 and 5") unless (1..5).include? level
+  end  
 
   def must_link_be_correct 
+    return unless video_link
     errors.add(:video_link,"This link is uncorrect") unless video_link =~ URI::regexp
   end
+
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:14]: remove this empty line

File: app/models/recipe.rb [hunk 3]
```diff
@@ -1,8 +1,15 @@
...
+  validate :must_link_be_correct,:must_level_be_correct
+  
+  def must_level_be_correct
+    return unless level
+    errors.add(:level,"the number must be between 1 and 5") unless (1..5).include? level
+  end  
 
   def must_link_be_correct 
+    return unless video_link
     errors.add(:video_link,"This link is uncorrect") unless video_link =~ URI::regexp
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:12]:
```suggestion
    return unless video_link
    
    errors.add(:video_link,"This link is uncorrect") unless video_link =~ URI::regexp
```

besides you can use rails validation with the format.

File: app/serializers/recipe_serializer.rb [hunk 4]
```diff
@@ -1,5 +1,11 @@
 class RecipeSerializer < ActiveModel::Serializer
-  attributes :id, :name, :content, :video_link
+  attributes :id, :name, :content, :video_link, :level
+
+  def level
+    return object.level if object.level
+
+    "Unknow"
```
Comment (PatrykTyszka) [inline app/serializers/recipe_serializer.rb:7]: 1. Typo
2. use const
3. single quotes

Comment (PatrykTyszka) [reply app/serializers/recipe_serializer.rb:7]: besides
```
return UNKNOWN unless object.level.present?

object.level
```

File: app/controllers/api/v1/recipes_controller.rb [hunk 5]
```diff
@@ -15,30 +15,30 @@ def create
         if @recipe.save
           render json: @recipe
         else
-          render json: { errors: @recipe.errors.to_s }, status: :unprocessable_entity
+          render json: { errors: @recipe.errors.messages }, status: :unprocessable_entity
```
Comment (PatrykTyszka) [inline app/controllers/api/v1/recipes_controller.rb:18]: 👍

File: db/migrate/20210816113045_add_level_field_to_recipes.rb [hunk 6]
```diff
@@ -0,0 +1,5 @@
+class AddLevelFieldToRecipes < ActiveRecord::Migration[6.1]
+  def change
+    add_column :recipes, :level, :integer
```
Comment (PatrykTyszka) [inline db/migrate/20210816113045_add_level_field_to_recipes.rb:3]: 👍

**Review by PatrykTyszka — CHANGES_REQUESTED** [review]

File: app/models/recipe.rb (hunk 1 above)
Comment (MarKoz162) [reply app/models/recipe.rb:8]: When I use inclusion I got a erroor message when level field is empty istead of "Unknown".
That's why I use a function to check before validation if level is empty.

Comment (PatrykTyszka) [reply app/models/recipe.rb:8]: you can add `allow_blank: true` option.

Comment (PatrykTyszka) [reply app/models/recipe.rb:8]: first of all, please read documentation :)

File: app/models/recipe.rb [hunk 7]
```diff
@@ -1,8 +1,11 @@
 class Recipe < ApplicationRecord
   validates :name, presence: true
-  validate :must_link_be_correct 
-
+  validate :must_link_be_correct
+  validates :level, inclusion:{ in: 1..5 ,message: "the number must be between 1 and 5"}, allow_blank:true
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:4]: the message will be generated automatically, you don't need to add this here unless it's required

File: app/models/recipe.rb [hunk 8]
```diff
@@ -1,8 +1,11 @@
...
+  
   def must_link_be_correct 
+    return unless video_link
+
     errors.add(:video_link,"This link is uncorrect") unless video_link =~ URI::regexp
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:9]: how about the format validation?

File: app/serializers/recipe_serializer.rb [hunk 9]
```diff
@@ -1,9 +1,15 @@
...
+
+  def level
+    return 'UNKNOWN' unless object.level.present?
+
+    object.level
+  end  
   
   def video_link
-    return "UNKNOWN" if object.video_link.blank?
+    return 'UNKNOWN' if object.video_link.blank?
```
Comment (PatrykTyszka) [inline app/serializers/recipe_serializer.rb:11]: you should add this as CONSTANT. Uppercase is not a constant variable :D

File: app/serializers/recipe_serializer.rb [hunk 10]
```diff
@@ -1,9 +1,15 @@
...
 
     object.video_link
   end
-end
+end
```
Comment (PatrykTyszka) [inline app/serializers/recipe_serializer.rb:15]: add an empty line at the end of the file...

**Review by PatrykTyszka — CHANGES_REQUESTED** [review]

File: app/models/recipe.rb [hunk 11]
```diff
@@ -1,8 +1,6 @@
...
   validates :name, presence: true
-  validate :must_link_be_correct 
+  validates :video_link, format:{ with: URI::regexp, message:"This link is uncorrect"}, allow_blank:true
+  validates :level, inclusion:{ in: 1..5 }, allow_blank:true
 
-  def must_link_be_correct 
-    errors.add(:video_link,"This link is uncorrect") unless video_link =~ URI::regexp
-  end
-end
+end
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:6]: empty line

File: app/models/recipe.rb [hunk 12]
```diff
@@ -1,8 +1,6 @@
 class Recipe < ApplicationRecord
   validates :name, presence: true
-  validate :must_link_be_correct 
+  validates :video_link, format:{ with: URI::regexp, message:"This link is uncorrect"}, allow_blank:true
+  validates :level, inclusion:{ in: 1..5 }, allow_blank:true
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:4]:
```suggestion
  validates :level, inclusion:{ in: 1..5 }, allow_blank: true
```

File: app/models/recipe.rb [hunk 13]
```diff
@@ -1,8 +1,6 @@
 class Recipe < ApplicationRecord
   validates :name, presence: true
-  validate :must_link_be_correct 
+  validates :video_link, format:{ with: URI::regexp, message:"This link is uncorrect"}, allow_blank:true
```
Comment (PatrykTyszka) [inline app/models/recipe.rb:3]:
```suggestion
  validates :video_link, format:{ with: URI::regexp, message: "This link is incorrect" }, allow_blank: true
```

File: app/serializers/recipe_serializer.rb [hunk 14]
```diff
@@ -1,9 +1,16 @@
 class RecipeSerializer < ActiveModel::Serializer
-  attributes :id, :name, :content, :video_link
+  attributes :id, :name, :content, :video_link, :level
+  UNKNOWN = 'UNKNOWN'
```
Comment (PatrykTyszka) [inline app/serializers/recipe_serializer.rb:3]:
```suggestion
  UNKNOWN = 'UNKNOWN'
  
  attributes :id, :name, :content, :video_link, :level
```

File: app/serializers/recipe_serializer.rb (hunk 10 above)
Comment (PatrykTyszka) [reply app/serializers/recipe_serializer.rb:16]: empty line

**Review by PatrykTyszka — CHANGES_REQUESTED** [review]

**Review by h4b00 — APPROVED** [review]

**Pull request merged**